File: oss/clawpipe/server/ios-app/FinSavvyAI/Services/WebSocketManager.swift
//
//  WebSocketManager.swift
//  WebSocket client for real-time cluster updates
//

import Combine
import Foundation

/// Manages a WebSocket connection to the FinSavvyAI desktop backend
/// for real-time push notifications (node changes, config updates, etc.).
class WebSocketManager: NSObject, ObservableObject, URLSessionWebSocketDelegate {
    @Published var isConnected = false
    @Published var lastMessage: WebSocketEvent?

    private var webSocketTask: URLSessionWebSocketTask?
    private var session: URLSession?
    private var url: URL?
    private var reconnectAttempts = 0
    private let maxReconnectAttempts = 10
    private var reconnectTimer: Timer?

    /// Callback invoked on the main queue when a relevant event arrives.
    var onEvent: ((WebSocketEvent) -> Void)?

    override init() {
        super.init()
    }

    // MARK: - Connection

    func connect(to urlString: String) {
        disconnect()

        guard let url = URL(string: urlString) else { return }
        self.url = url

        let config = URLSessionConfiguration.default
        session = URLSession(configuration: config, delegate: self, delegateQueue: .main)
        webSocketTask = session?.webSocketTask(with: url)
        webSocketTask?.resume()

        receiveMessage()
        startPing()
    }

    func disconnect() {
        reconnectTimer?.invalidate()
        reconnectTimer = nil
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        session?.invalidateAndCancel()
        session = nil

        DispatchQueue.main.async {
            self.isConnected = false
        }
    }

    // MARK: - Messaging

    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            guard let self = self else { return }

            switch result {
            case .success(let message):
                self.handleMessage(message)
                self.receiveMessage()  // Continue listening
            case .failure(let error):
                print("WebSocket receive error: \(error.localizedDescription)")
                self.handleDisconnect()
            }
        }
    }

    private func handleMessage(_ message: URLSessionWebSocketTask.Message) {
        var text: String?
        switch message {
        case .string(let str):
            text = str
        case .data(let data):
            text = String(data: data, encoding: .utf8)
        @unknown default:
            break
        }

        guard let text = text,
            let data = text.data(using: .utf8),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let eventType = json["type"] as? String
        else { return }

        let event = WebSocketEvent(
            type: eventType,
            payload: json["data"]
        )

        DispatchQueue.main.async {
            self.lastMessage = event
            self.onEvent?(event)
        }
    }

    // MARK: - Ping / Keep-alive

    private func startPing() {
        webSocketTask?.sendPing { [weak self] error in
            guard let self = self, error == nil else { return }
            DispatchQueue.main.asyncAfter(deadline: .now() + 30) {
                self.startPing()
            }
        }
    }

    // MARK: - Reconnection with Exponential Backoff

    private func handleDisconnect() {
        DispatchQueue.main.async {
            self.isConnected = false
        }

        guard reconnectAttempts < maxReconnectAttempts else {
            print("WebSocket: max reconnect attempts reached")
            return
        }

        let delay = min(pow(2.0, Double(reconnectAttempts)), 30.0)  // 1, 2, 4, 8, ... 30s max
        reconnectAttempts += 1

        print("WebSocket: reconnecting in \(delay)s (attempt \(reconnectAttempts))")

        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self = self, let url = self.url else { return }
            self.connect(to: url.absoluteString)
        }
    }

    // MARK: - URLSessionWebSocketDelegate

    func urlSession(
        _ session: URLSession, webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol: String?
    ) {
        DispatchQueue.main.async {
            self.isConnected = true
            self.reconnectAttempts = 0
        }
    }

    func urlSession(
        _ session: URLSession, webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?
    ) {
        handleDisconnect()
    }
}

// MARK: - Event Model

struct WebSocketEvent: Identifiable {
    let id = UUID()
    let type: String
    let payload: Any?
    let timestamp = Date()

    /// Common event types from the Go backend
    var isNodeEvent: Bool {
        type.hasPrefix("node_")
    }

    var isClusterEvent: Bool {
        type.hasPrefix("cluster_")
    }

    var isConfigEvent: Bool {
        type == "config_updated"
    }
}

File: products/queryflux/web/lib/websocket-manager.ts
/**
 * WebSocket Manager — handles real-time connection with auto-reconnect
 */

export class WebSocketManager {
  private ws: WebSocket | null = null;
  private url: string;
  private token: string | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private messageHandlers: Map<string, (data: any) => void> = new Map();
  private isIntentionallyClosed = false;

  constructor(url: string, token: string | null = null) {
    this.url = url;
    this.token = token;
  }

  public connect(): void {
    if (this.ws?.readyState === WebSocket.OPEN) return;
    try {
      const wsUrl = this.token ? `${this.url}?token=${this.token}` : this.url;
      this.ws = new WebSocket(wsUrl);

      this.ws.onopen = () => {
        console.log('WebSocket connected');
        this.reconnectAttempts = 0;
      };

      this.ws.onmessage = (event) => {
        try {
          this.handleMessage(JSON.parse(event.data));
        } catch (error) {
          console.error('Failed to parse WebSocket message:', error);
        }
      };

      this.ws.onerror = (error) => { console.error('WebSocket error:', error); };

      this.ws.onclose = () => {
        console.log('WebSocket closed');
        if (!this.isIntentionallyClosed && this.reconnectAttempts < this.maxReconnectAttempts) {
          this.reconnect();
        }
      };
    } catch (error) {
      console.error('Failed to create WebSocket connection:', error);
    }
  }

  public disconnect(): void {
    this.isIntentionallyClosed = true;
    if (this.ws) { this.ws.close(); this.ws = null; }
  }

  private reconnect(): void {
    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);
    console.log(`Reconnecting WebSocket in ${delay}ms (attempt ${this.reconnectAttempts}/${this.maxReconnectAttempts})`);
    setTimeout(() => this.connect(), delay);
  }

  public send(type: string, data: any): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({ type, data }));
    } else {
      console.warn('WebSocket is not connected. Message not sent:', { type, data });
    }
  }

  public on(type: string, handler: (data: any) => void): void {
    this.messageHandlers.set(type, handler);
  }

  public off(type: string): void {
    this.messageHandlers.delete(type);
  }

  private handleMessage(message: { type: string; data: any }): void {
    const handler = this.messageHandlers.get(message.type);
    if (handler) {
      handler(message.data);
    } else {
      console.warn('No handler for WebSocket message type:', message.type);
    }
  }

  public updateToken(token: string): void {
    this.token = token;
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.disconnect();
      this.isIntentionallyClosed = false;
      this.connect();
    }
  }
}
